### DISCUSSION TITLE
set/get methods of Model Class

### DISCUSSION THREAD

**Discussion by solidkon** [mupif/mupif#65] [Q&A]
Hi!
I'm currently in the process of using the get/set methods of the DemoModel() class provided by @stanislavsulc on [https://gist.github.com/stanislavsulc/5944f1f10d50f37ece05690b8a4d9da7 ](url) These set/get methods only assume that the Inputs of the model are of type mupif.Heavystruct. Of course for the purpose of the DemoModel() provided this is sufficient and pretty much clear. However, now that I'm building the actual API of the model I need to know the **_general rules_** for defining every different possible Input for our models. If I'm not missing anything, possible inputs for our models would be:

1. mupif.Heavystruct
2. mupif.PyroFile
3. mupif.Property
(Probably a spatial field would not be relevant for discrete electronic/atomistic/mesoscopic simulations but I cannot tell with **absolute certainty at the moment**)

I see that the args of the get method are: objectTypeID, while objectID is a kwarg. Now that @eudoxos suggested as temporary solution for arbitrary number of heavystructs and pyrofiles on [https://github.com/mupif/mupif/discussions/56](url), I'm not sure how to pass this properly (via set) and get it accordingly. It would also be great if there is some sort of documentation on the general rules and good practices for defining a Model API correctly.

**Comment by stanislavsulc** [comment]
Hi,
the set method is defined as
`def set(self, obj, objectID=""):`
It can check obj with
`if obj.isInstance(mp.PyroFile):`
or
`if obj.isInstance(mp.Property):`
and then work with it according to the objectID, which represents the number of the particular input as a string.

**Reply by solidkon** [reply]
Thank you stanislav! I have already defined it in this way but I'm getting:

`AttributeError: 'PyroFile' object has no attribute 'id'`

I attach the set method:

```
    def set(self, obj, objectID=""):
        if obj.isInstance(mp.HeavyStruct):
            data_id = "mupif." + str(obj.id)
            self._setInput(obj, data_id, objectID)
        elif obj.isInstance(mp.PyroFile):
            data_id = "mupif." + str(obj.id)
            self._setInput(obj, data_id, objectID)
        elif obj.isInstance(mp.Property):
            data_id = "mupif." + str(obj.id)
            self._setInput(obj, data_id, objectID)
```

**Reply by stanislavsulc** [reply]
HeavyStruct has the id attribute, which represents the value from DataID enum, but other objects do not. The DataID value can be obtained from Property with getPropertyID() method. But PyroFile has no DataID attribute.

**Reply by stanislavsulc** [reply]
See the set and get methods in this https://github.com/mupif/mupif/blob/master/examples/models.py or other mupif examples: 
Part of the code is here:
https://github.com/mupif/mupif/blob/53a8192cbba0a28b9cb24da49f74eeea6aba07e9/examples/models.py#L598-L625

**Reply by solidkon** [reply]
Thank you. I will try to adjust the relevant fields. Just a remark; it would be much easier if the examples you share are implemented on the concept of **indirect API implementation** and not the native one, since it's the most common case for the MUSICODE project and more traceable (abstracted from the solver physics and operations).

**Reply by solidkon** [reply]
No it's not clear how a mupif.Pyrofile will be assigned to Inputs via the `_setInput(self, val, data_id, obj_id)` method of the class, which has a non-optional data_id argument.

**Reply by stanislavsulc** [reply]
You can just use the parameter "mupif.DataID.ID_InputFile" for all inputs of type PyroFile.
```
if obj.isInstance(mp.PyroFile):
    self._setInput(obj, "mupif.DataID.ID_InputFile", objectID)
```
